contracts/contracts/ccip/merkle_root.tolk

```diff
@@ -26,36 +28,67 @@ fun MerkleRoot_Storage.store(self) {
     return contract.setData(self.toCell());
 }
 
-type Msg = ExecuteReport;
+
+type Msg = MerkleRoot_Validate;
 
 fun onInternalMessage(in: InMessage) {
     val msg = lazy Msg.fromSlice(in.body);
     match (msg) {
-        ExecuteReport => {}
+        MerkleRoot_Validate => {
+            // If we receive this message from the OffRamp then the merkle proof has already been validated.
+            _validateAndExecute(msg, in.senderAddress);
+            
+        }
         else => {
             // ignore empty messages, "wrong opcode" for others
             assert (in.body.isEmpty()) throw 0xFFFF
         }
     }
 }
 
-// fun onExecute() { // TODO: rename to something like process
-    // validate we trust the onramp (is owner)
+//TODO: We are not keeping track of the execution state right now, we just relay the message to confirm that the merkle proof is valid
+
+fun _validateAndExecute(msg: MerkleRoot_Validate, sender: address) {
+    val st = MerkleRoot_Storage.load();
+    assert(sender == st.owner, ERROR_NOT_OWNER);
+
+    assert(st.state == EXECUTION_STATE_UNTOUCHED, ERROR_STATE_IS_NOT_UNTOUCHED);
+    // TODO if no tokenTransfers
+    _execute(msg.messages, msg.proofs, msg.proofFlagBits, msg.metadataHash, sender);
+
+    // if tokenTransfers 
     // validate execution state == untouched
-    //
     // if initial
     //  allocate storage space for the execution report
     // TODO: should we fan out CommitReport into individual executions?
     // 
     // if in progress
-    // 
     // if tokenAmounts > 0 && state == initial
     // set state = token transfer
     // send message to offramp to ask to release tokens
     // else if no tokenTransfers
     // set state = execution
-    // call execute
-// }
+}
+
+fun _execute(mesages: cell, proofs: cell, proofFlagBits: uint256, metadataHash: uint256, offRamp: address) {
+    var st = MerkleRoot_Storage.load();
```

Comment:
Consider extracting contract state loading and (finally) storing to the external handler, so it can manage storage is loaded once, processed, and finally stored.

Here the flow loads storage twice.